### PR TITLE
Add AArch64 memory operations (LDR/STR/LDP/STP) — closes #68

### PR DESCRIPTION
**Pull request by pmatos**
## Summary
- Adds the full AArch64 memory-op family — LDR, LDRB, LDRH, LDRSB, LDRSH, LDRSW, STR, STRB, STRH, LDP, STP, LDPSW — across all five addressing modes (immediate-offset, pre-index, post-index, register-offset, register-extend) and both W/X widths.
- End-to-end pipeline: IR → parser → concrete interpreter → SMT (Z3 Array) → equivalence checker → assembler → cost → enumerative/stochastic search candidates → Capstone-tripwire tests → ELF round-trip integration tests.
- Sound aliasing via a single byte-addressed `Array<BV64, BV8>` on the SMT side and a sparse `BTreeMap<u64, u8>` (absent-key = 0, prune-on-write) on the concrete side; little-endian by construction. See [ADR-0007](docs/adr/0007-memory-model.md).

## Design decisions (pinned in ADR-0007)
- **Memory model**: byte-addressed Z3 array (SMT) + sparse `BTreeMap` (concrete). Sound full aliasing — no disjointness preconditions.
- **Live-out**: whole-memory live-out is auto-derived from `touches_memory()` on the candidate / target; no CLI grammar change.
- **`fast_only` carve-out**: force-disabled inside `check_equivalence_with_config` for memory-bearing windows (random inputs can't cover all aliasing patterns).
- **Writeback** via the broadened `Instruction::destinations() -> Vec<Register>` (pre-/post-index appends the base register).
- **LDPSW** folded into `Ldp` with a `signed: bool` field; encodable only when `width == Word`.
- **`Xt == Xn` writeback** and **LDP `Xt == Xt2`** rejected at `is_encodable_aarch64` per ARM ARM.

## Commit ledger (14 commits)
| # | Commit | What |
|---|--------|------|
| 1 | `a454ae0` | docs: ADR-0007 + plan |
| 2 | `0d81dea` | ir: `AddressOperand` / `IndexMode` / `AccessWidth` |
| 3 | `9aed269` | ir: variants + `destinations()` API |
| 4 | `d89867b` | parser: all addressing modes, bracket-aware tokenizer |
| 5 | `485bdc1` | semantics: sparse memory on `ConcreteMachineState` |
| 6 | `759bf81` | semantics: concrete eval for LDR/LDRS/STR/LDP/STP |
| 7 | `16da2ae` | semantics: SMT lowering via Z3 Array |
| 8/9 | `b579919` | semantics: auto-derive `memory_live` + `fast_only` carve-out |
| 17 | `51169cd` | test: extend Capstone tripwire with 54 memory-op rows |
| 21 | `755a410` | docs: update CLAUDE.md |
| 13 | `e873441` | assembler: dynasm encoder with LDUR/STUR dispatch (29 round-trip tests) |
| 11 | `366d576` | random: memory seed for memory-bearing windows |
| 20 | `7d07aee` | test: 3 end-to-end opt-window tests (STP/LDP/LDR in `loops_debug`) |
| 15/16 | `82c8888` | search: enumerative candidates + stochastic mutator slots |

Every commit passes `./ci_check.sh` independently.

## Out-of-scope follow-ups
- W-form enumeration (mutator can still flip width via a future opcode bridge).
- LDUR/STUR mnemonics as a parser entry point (round-trip via Capstone today produces them when the encoder routes a negative-offset LDR to LDUR; the parser pins them as `Unsupported` to avoid silent semantic drift).
- `LiveOutMask<R>` migration (ADR-0004 §5; ADR-0007 notes the `memory_live` bit moves there).
- Mutate-opcode width / signed bridges (e.g. `Ldr ↔ Ldrs`).

## Test plan
- [x] `./ci_check.sh` (formatting, build, unit tests, integration tests)
- [x] All 116 assembler tests pass (29 new memory-op round-trips)
- [x] Capstone-tripwire pins all 12 new mnemonics × 5 addressing modes (54 new rows)
- [x] 3 integration tests run the full `s11 opt` pipeline on real ELF windows containing STP / LDP / LDR
- [x] `mutate_operand_on_ldr_can_change_base_or_rt_or_offset` proves the mutator touches operand fields without changing variant/width/mode
- [x] `generate_all_instructions_includes_memory_ops` proves the enumerative pool produces each of LDR/STR/LDP/STP

Closes #68.

### CODE REVIEW

File: src/assembler/mod.rs [hunk 1]
```diff
@@ -292,6 +294,302 @@ macro_rules! emit_ccmp_imm {
...
+                } else {
+                    dynasm!($ops ; .arch aarch64 ; $mnem $rt_tok(rt_n), [XSP(base_n), X(idx_n), SXTX shift_n]);
+                }
+                Ok::<(), String>(())
+            }
+            _ => Err(format!(
+                "{} cannot use extend kind {:?} (rejected by is_encodable_aarch64)",
+                stringify!($mnem),
+                $kind
+            )),
```
[the remark this answers is not in the document]
Comment (pmatos) [reply src/assembler/mod.rs:421]: Fixed in `e477f0c`.

Changed: added a `ExtendKind::Uxtx` arm to `emit_mem_ext!` in `src/assembler/mod.rs`. UXTX on an X-form index is architecturally equivalent to LSL (both select option=011 in the LDR-register encoding), and dynasm-rs's memory-operand sanitiser disallows the `UXTX` token but accepts the bare/`LSL` form — so the new arm dispatches through that path, mirroring how GNU `as` rewrites `[Xn, Xm, UXTX]` to `[Xn, Xm]`.

Validation: two new round-trip tests `ldr_x_uxtx_extended_index_no_shift` and `ldr_x_uxtx_extended_index_shift_three` pin the Capstone disassembly (`[x1, x2]` and `[x1, x2, lsl #3]` respectively).

_This assessment was made by two independent AI reviewers (Claude Opus 4.7 and Codex GPT-5.5). If you disagree, please reply and we'll re-evaluate._

File: src/ir/instructions.rs [hunk 2]
```diff
@@ -940,8 +1100,123 @@ impl Instruction {
...
+        return false;
+    }
+    // Writeback `base == rtN` is rejected.
+    if is_writeback(addr) {
+        let base = address_base(addr);
+        if base == rt1 || base == rt2 {
+            return false;
         }
     }
+    true
```
[the remark this answers is not in the document]
Comment (pmatos) [reply src/ir/instructions.rs:1238]: Fixed in `6a2c43a`.

Changed: `is_encodable_pair` in `src/ir/instructions.rs` now explicitly rejects `AddressOperand::Reg` / `Ext` (pair forms only accept immediate addressing modes per ARM ARM C6.2 / C6.2.123 / C6.2.205), so `parse_line`'s `is_encodable_aarch64` final gate catches the mismatch before the assembler is asked to emit it. The same edit also added a 7-bit signed scaled immediate range check (folding in claude[bot]'s related review item #8).

Validation: 5 new tests in `src/ir/instructions.rs` — `pair_reg_offset_mode_rejected_at_encodability`, `pair_ext_mode_rejected_at_encodability`, `pair_imm_offset_out_of_range_rejected`, `pair_imm_offset_unscaled_rejected`, `pair_imm_offset_in_range_accepted`.

_This assessment was made by two independent AI reviewers (Claude Opus 4.7 and Codex GPT-5.5). If you disagree, please reply and we'll re-evaluate._

File: src/search/stochastic/mutation.rs [hunk 3]
```diff
@@ -1098,6 +1200,75 @@
...
+            addr: AddressOperand::Imm {
+                base: Register::X1,
+                offset: 0,
+                mode: IndexMode::Offset,
+            },
+            width: AccessWidth::Extended,
+        };
+        let mut changed = false;
+        for _ in 0..200 {
+            let mut seq = vec![initial.clone()];
```
[the remark this answers is not in the document]
Comment (pmatos) [reply src/search/stochastic/mutation.rs]: Fixed in `af04889`. Dropped the `.clone()` — `Instruction` is `Copy`, so the binding can use `vec![initial]` directly.

File: src/semantics/smt.rs [hunk 4]
```diff
@@ -203,6 +220,106 @@
...
+                    IndexMode::PostIndex => (base_bv, Some((*base, updated))),
+                }
+            }
+            AddressOperand::Reg { base, idx, shift } => {
+                let base_bv = self.get_register(*base).clone();
+                let idx_bv = self.get_register(*idx).clone();
+                let shifted = if *shift == 0 {
+                    idx_bv
+                } else {
+                    idx_bv.bvshl(&BV::from_u64(*shift as u64, 64))
```
[the remark this answers is not in the document]
Comment (pmatos) [reply src/semantics/smt.rs]: Fixed in `af04889`. Removed the leading `&` on `BV::from_u64(...)` — `bvshl` accepts an owned `BV` value.

[the remark this answers is not in the document]
Comment (pmatos) [reply src/semantics/smt.rs]: Fixed in `af04889`. Same `needless_borrows_for_generic_args` fix as the sibling thread on this line — passing the `BV` by value to `bvshl`.

File: src/semantics/smt.rs [hunk 5]
```diff
@@ -203,6 +220,106 @@
...
+                    // keep apply_instruction total.
+                    ExtendKind::Uxtb => idx_bv.extract(7, 0).zero_ext(56),
+                    ExtendKind::Uxth => idx_bv.extract(15, 0).zero_ext(48),
+                    ExtendKind::Sxtb => idx_bv.extract(7, 0).sign_ext(56),
+                    ExtendKind::Sxth => idx_bv.extract(15, 0).sign_ext(48),
+                };
+                let shifted = if *shift == 0 {
+                    extended
+                } else {
+                    extended.bvshl(&BV::from_u64(*shift as u64, 64))
```
[the remark this answers is not in the document]
Comment (pmatos) [reply src/semantics/smt.rs]: Fixed in `af04889`. Same `needless_borrows_for_generic_args` fix as the smt.rs:289 sibling — passing the `BV` by value to `bvshl` in the Ext-mode address path.

[the remark this answers is not in the document]
Comment (pmatos) [reply src/semantics/smt.rs]: Fixed in `af04889` (duplicate of the sibling thread on this line).

File: src/semantics/smt.rs [hunk 6]
```diff
@@ -857,10 +974,113 @@
...
+            let extended = ldr_sign_extend(&raw, *width, state.width);
+            state.set_register(*rt, extended);
+            if let Some((base, new_base)) = writeback {
+                state.set_register(base, new_base);
+            }
+        }
+        Instruction::Str { rt, addr, width } => {
+            let (effective, writeback) = state.eval_address(addr);
+            let value = state.get_register(*rt).clone();
+            let bits = (width.bytes() * 8) as u32;
```
[the remark this answers is not in the document]
Comment (pmatos) [reply src/semantics/smt.rs]: Fixed in `af04889`. Dropped the redundant `as u32` — `width.bytes() * 8` is already `u32`.

[the remark this answers is not in the document]
Comment (pmatos) [reply src/semantics/smt.rs]: Fixed in `af04889` (duplicate of the sibling thread on this line).

File: src/semantics/smt.rs [hunk 7]
```diff
@@ -857,10 +974,113 @@
...
+        }
+        Instruction::Stp {
+            rt1,
+            rt2,
+            addr,
+            width,
+        } => {
+            let (effective, writeback) = state.eval_address(addr);
+            let bytes = width.bytes();
+            let bits = (bytes * 8) as u32;
```
[the remark this answers is not in the document]
Comment (pmatos) [reply src/semantics/smt.rs]: Fixed in `af04889`. Dropped the redundant `as u32` — `bytes * 8` is already `u32`.

[the remark this answers is not in the document]
Comment (pmatos) [reply src/semantics/smt.rs]: Fixed in `af04889` (duplicate of the sibling thread on this line).

File: src/semantics/smt.rs [hunk 8]
```diff
@@ -857,10 +974,113 @@
...
+                state.set_register(base, new_base);
+            }
+        }
     }
     state
 }
 
+/// Zero-extend the low `width.bytes() * 8` bits of `raw` to `target_width`.
+fn ldr_zero_extend(raw: &BV, width: AccessWidth, target_width: u32) -> BV {
+    let raw_bits = (width.bytes() * 8) as u32;
```
[the remark this answers is not in the document]
Comment (pmatos) [reply src/semantics/smt.rs]: Fixed in `af04889`. Dropped the redundant `as u32` in `ldr_zero_extend` — `width.bytes() * 8` is already `u32`.

[the remark this answers is not in the document]
Comment (pmatos) [reply src/semantics/smt.rs]: Fixed in `af04889` (duplicate of the sibling thread on this line).

File: src/semantics/smt.rs [hunk 9]
```diff
@@ -857,10 +974,113 @@
...
+    if pad == 0 {
+        raw.clone()
+    } else {
+        raw.zero_ext(pad)
+    }
+}
+
+/// Sign-extend the low `width.bytes() * 8` bits of `raw` to `target_width`.
+fn ldr_sign_extend(raw: &BV, width: AccessWidth, target_width: u32) -> BV {
+    let raw_bits = (width.bytes() * 8) as u32;
```
[the remark this answers is not in the document]
Comment (pmatos) [reply src/semantics/smt.rs]: Fixed in `af04889`. Dropped the redundant `as u32` in `ldr_sign_extend` — `width.bytes() * 8` is already `u32`.

[the remark this answers is not in the document]
Comment (pmatos) [reply src/semantics/smt.rs]: Fixed in `af04889` (duplicate of the sibling thread on this line).

File: src/semantics/smt.rs [hunk 10]
```diff
@@ -3215,4 +3442,103 @@
...
+                width: AccessWidth::Extended,
+            },
+        ];
+        let post = apply_sequence(pre.clone(), &seq);
+        // The Z3 solver must be able to verify the load returns the expected
+        // mixed bytes. We just check satisfiability of a concrete example.
+        let solver = Solver::new();
+        let x0_pre = pre.get_register(Register::X0);
+        let zero = BV::from_u64(0, 64);
+        solver.assert(x0_pre.eq(&BV::from_u64(0xDEADBEEF_CAFEBABE, 64)));
```
[the remark this answers is not in the document]
Comment (pmatos) [reply src/semantics/smt.rs]: Fixed in `af04889`. Passed the `BV::from_u64(...)` by value to `.eq()` instead of `&BV::from_u64(...)`.

File: src/semantics/smt.rs [hunk 11]
```diff
@@ -3215,4 +3442,103 @@
...
+        solver.assert(pre.get_register(Register::X1).eq(&zero));
+        // Expected: low 2 bytes of x2 from x0 low 2 bytes = 0xBABE
+        // Then bytes 2..3 from x0 low 2 bytes (re-stored) = 0xBABE
+        // Then bytes 4..7 from x0 high 4 bytes = 0xDEADBEEF
+        let x2_post = post.get_register(Register::X2);
+        let expected = BV::from_u64(0xDEADBEEF_BABE_BABE, 64);
```
[the remark this answers is not in the document]
Comment (pmatos) [reply src/semantics/smt.rs]: Fixed in `af04889`. Regrouped the hex literal `0xDEADBEEF_BABE_BABE` to `0xDEAD_BEEF_BABE_BABE` (consistent 4-hex-digit groups).

File: src/validation/random.rs [hunk 12]
```diff
@@ -223,6 +252,46 @@
...
+        // At least one input must have a non-empty memory map (the random
+        // bytes might all be zero, but with 4 KiB of random bytes and four
+        // independent draws the probability of every byte being zero is
+        // 256^-(4*4096) — vastly below 2^-256, indistinguishable from zero).
+        let any_populated = inputs.iter().any(|s| !s.memory().is_empty());
+        assert!(any_populated, "memory seed produced no entries");
+        // Every populated byte must lie inside [MEMORY_SEED_BASE,
+        // MEMORY_SEED_BASE + MEMORY_SEED_SIZE).
+        for s in &inputs {
+            for (&addr, _) in s.memory() {
```
[the remark this answers is not in the document]
Comment (pmatos) [reply src/validation/random.rs]: Fixed in `af04889`. Switched the inner loop from `for (&addr, _) in s.memory()` to `for &addr in s.memory().keys()`.

File: tests/integration/opt_test.rs [hunk 13]
```diff
@@ -448,6 +449,113 @@
     );
 }
 
+/// Helper for memory-op integration tests: assert that `s11 opt` on the
+/// given single-instruction window succeeds.
+///
+/// Each test copies the source ELF to a unique tmp path (`std::env::temp_dir`
+/// + the supplied `fixture_tag`) so concurrent `cargo test` runs don't
+/// collide on the `<input>_optimized` artifact the binary always writes
```
[the remark this answers is not in the document]
Comment (pmatos) [reply tests/integration/opt_test.rs]: Fixed in `af04889`. Rewrote the docstring to avoid the leading `+` that clippy interpreted as a Markdown list bullet (`+ the supplied` → `(... joined with the supplied ...)`).

File: tests/integration/opt_test.rs [hunk 14]
```diff
@@ -448,6 +449,113 @@
...
+/// alongside its input. The artifact is cleaned up before the test returns.
```
[the remark this answers is not in the document]
Comment (pmatos) [reply tests/integration/opt_test.rs]: Fixed in `af04889` (same docstring rewrite as the sibling thread on line 457).

**Comment by pmatos** [comment]
@claude Regarding your code review (https://github.com/pmatos/s11/pull/289#issuecomment-4481020099):

Each of the 12 items was independently triaged by two AI reviewers (Claude Opus 4.7 and Codex GPT-5.5). Where both agreed FIX, the change landed in this PR; where both agreed REJECT, no code change was made; otherwise the item was deferred to a tracking issue. Per-item outcome:

| # | Outcome | Action |
|---|---------|--------|
| 1 — `states_not_equal` async w/ memory | **REJECT** (not-an-issue) | The unconditional `states_not_equal` is the unmasked full-state helper; `states_not_equal_for_live_out` correctly gates on `memory_live`. The asymmetry is intentional. |
| 2 — `ldr_zero_extend` underflow | **DEFER** (diminishing-returns) | Underflow is unreachable today (`is_encodable_aarch64` rejects `Ldrs Extended`). Tracked as #290. |
| 3 — `compute_address` `Reg` index width | **REJECT** (not-an-issue) | Full 64-bit X-register index is architecturally correct for the `Reg` variant. |
| 4 — Test comment misleading | **DEFER** (diminishing-returns) | Value correct, comment polishing only. Tracked as #291. |
| 5 — Redundant `#[allow(dead_code)]` | **DEFER** (diminishing-returns) | Cosmetic cleanup. Tracked as #292. |
| 6 — `ldrs_mnemonic` Extended fallthrough | **DEFER** (diminishing-returns) | Unreachable via the encodability gate. Tracked as #293. |
| 7 — `opcode_id` duplication | **DEFER** (scope-creep) | Predates this PR; consolidation is a separate refactor. Tracked as #294. |
| 8 — LDP/STP imm-range missing | **FIX** | Fixed in `6a2c43a` (folded into `is_encodable_pair` alongside codex's Reg/Ext rejection). 5 new tests pin the boundary cases. |
| 9 — SMT memory equality cost | **REJECT** (not-an-issue) | `build_smt_solver` uses `states_not_equal_for_live_out` with `config.memory_live`, so the term is gated, not unconditional. |
| 10 — `parse_lsl_amount` over-restricts | **DEFER** (diminishing-returns) | Parser accepts, `is_encodable_aarch64` rejects — behaviourally sound, UX nit. Tracked as #295. |
| 11 — `random_address_offset` `-256` | **DEFER** (diminishing-returns) | Legal for LDUR/STUR & pre/post-index; search-quality tuning. Tracked as #296. |
| 12 — `Ldp Byte/Half` not blocked at ctor | **DEFER** (scope-creep) | Encodability gate handles it; construction-time prevention is a wider IR refactor. Tracked as #297. |

_This assessment was made by two independent AI reviewers (Claude Opus 4.7 and Codex GPT-5.5). If you disagree on any of the REJECTs, please reply and we'll re-evaluate._

File: src/ir/instructions.rs [hunk 15]
```diff
@@ -940,8 +1100,123 @@ impl Instruction {
...
+
+/// Encodability gate for the LDR / LDRS / STR family. Rules per ADR-0007:
+/// (a) base register cannot be XZR (the XSP slot rejects the zero register
+/// — SP is accepted); (b) `rt` cannot be SP (loads/stores use Xt/Wt
+/// slots); (c) writeback `rt == base` is CONSTRAINED UNPREDICTABLE and
+/// rejected at the IR level; (d) signed loads only support B/H/W access
+/// widths (LDRSX does not exist — LDR handles 64-bit zero-extension).
+fn is_encodable_ldr_like(rt: Register, addr: &AddressOperand, width: AccessWidth) -> bool {
+    if rt == Register::XZR || rt == Register::SP {
+        return false;
```
[the remark this answers is not in the document]
Comment (pmatos) [reply src/ir/instructions.rs]: Fixed in `cad3229`.

Changed: `is_encodable_ldr_like` in `src/ir/instructions.rs` now allows `rt == XZR` — `str xzr, [x0]` and `ldr xzr, [x0]` are legal per ARM ARM C6.2.131 / C6.2.205 (the load to ZR is discarded; the store writes zero). The XZR catch-all was previously dropping legal encodings.

Validation: 2 new tests in `src/ir/instructions.rs` — `str_xzr_is_encodable` and `ldr_xzr_is_encodable` — pin both directions.

_This assessment was made by two independent AI reviewers (Claude Opus 4.7 and Codex GPT-5.5). If you disagree, please reply and we'll re-evaluate._

File: src/ir/instructions.rs [hunk 16]
```diff
@@ -940,10 +1100,149 @@ impl Instruction {
...
+    }
+    if address_base(addr) == Register::XZR {
+        return false;
+    }
+    if is_writeback(addr) && rt == address_base(addr) {
```
[the remark this answers is not in the document]
Comment (pmatos) [reply src/ir/instructions.rs:1157]: Fixed in `cad3229`.

Changed: `is_encodable_ldr_like` now rejects `AddressOperand::Reg` / `AddressOperand::Ext` whose `idx == SP`. The Rm field encodes X0..X30 / XZR via `register_to_dynasm`; SP is only addressable through the XSP base slot, so the assembler errors otherwise. The IR gate now matches the assembler's contract before the candidate reaches emission.

Validation: 2 new tests in `src/ir/instructions.rs` — `ldr_with_sp_index_rejected` and `ldr_with_sp_index_via_ext_mode_rejected`.

_This assessment was made by two independent AI reviewers (Claude Opus 4.7 and Codex GPT-5.5). If you disagree, please reply and we'll re-evaluate._

File: src/parser/mod.rs [hunk 17]
```diff
@@ -711,6 +729,293 @@ fn is_extend_keyword(kw: &str) -> bool {
...
+fn parse_lsl_amount(tail: &str) -> Result<u8, String> {
+    let mut parts = tail.split_whitespace();
+    let kw = parts.next().unwrap_or("");
+    if !kw.eq_ignore_ascii_case("lsl") {
+        return Err(format!("expected `lsl`, got `{}`", kw));
+    }
+    let imm_tok = parts.next().ok_or("missing LSL amount")?;
+    let amt = parse_immediate(imm_tok)?;
+    if !(0..=63).contains(&amt) {
+        return Err(format!("LSL amount {} out of range", amt));
```
[the remark this answers is not in the document]
Comment (pmatos) [reply src/parser/mod.rs:891]: **Deferred (diminishing returns)** — Same concern as claude[bot] review item 10 ([triage table](https://github.com/pmatos/s11/pull/289#issuecomment-4481486812)). Current behaviour is sound: parser accepts `0..=63` because the same helper drives shifted-register ALU operands; `is_encodable_aarch64` rejects out-of-set shifts before SMT or emission. The change is a parse-time UX improvement (clearer error site). Tracked as #295.

_This assessment was made by two independent AI reviewers (Claude Opus 4.7 and Codex GPT-5.5). If you disagree, please reply and we'll re-evaluate._

File: src/parser/mod.rs [hunk 18]
```diff
@@ -711,6 +729,293 @@ fn is_extend_keyword(kw: &str) -> bool {
...
+        "sxtx" => ExtendKind::Sxtx,
+        _ => return Err(format!("unsupported memory extend keyword `{}`", kw)),
+    };
+    let shift = match parts.next() {
+        None => 0u8,
+        Some(imm_tok) => {
+            let amt = parse_immediate(imm_tok)?;
+            if !(0..=4).contains(&amt) {
+                return Err(format!("extend shift {} out of range (0..=4)", amt));
+            }
```
[the remark this answers is not in the document]
Comment (pmatos) [reply src/parser/mod.rs:915]: **Deferred (diminishing returns)** — Sibling concern to thread [r3261428046](https://github.com/pmatos/s11/pull/289#discussion_r3261428046) (LSL set) and claude[bot] review item 10. Current behaviour is sound on the `Ext` paths the assembler accepts (illegal extend-shifts hit the assembler error path); the change would tighten parser-time UX plus a per-width check in `is_encodable_ldr_like`. Tracked as #302.

_This assessment was made by two independent AI reviewers (Claude Opus 4.7 and Codex GPT-5.5). If you disagree, please reply and we'll re-evaluate._

**Comment by pmatos** [comment]
@claude Regarding your second code review ([comment 4481490724](https://github.com/pmatos/s11/pull/289#issuecomment-4481490724)):

The 7 items + 3 nits overlap heavily with your first review (already triaged in [the original table](https://github.com/pmatos/s11/pull/289#issuecomment-4481486812)). Per-item triage:

| # | Topic | Outcome |
|---|-------|---------|
| 1 | `states_not_equal` unconditional memory clause | **REJECT** (not-an-issue) — confirmed in [your own follow-up](https://github.com/pmatos/s11/pull/289#issuecomment-4481489080). |
| 2 | Missing `eprintln!` warning when `fast_only` is overridden | **DEFER** (diminishing-returns) — tracked as #303. |
| 3 | Duplicated `memory_touched` auto-derive in two functions | **REJECT** (already-handled) — resolved during this rebase: `check_equivalence_with_config` is now a thin wrapper around `check_equivalence_with_config_metrics`, so the block lives in exactly one place on the rebased tree (see `cad3229`). |
| 4 | `touches_memory` duplicates `has_side_effects` match arm | **DEFER** (diminishing-returns) — tracked as #304. |
| 5 | `ldrs_mnemonic(Extended)` should be `unreachable!()` | **DEFER** (already filed) — same as first review #6, tracked as #293. |
| 6 | Spurious `#[allow(dead_code)]` on `destinations()` helpers | **DEFER** (already filed) — same as first review #5, tracked as #292. |
| 7 | `opcode_id` duplicated between aarch64.rs and candidate.rs | **DEFER** (already filed) — same as first review #7, tracked as #294. |
| Nit | `[Xn, #0], #imm` post-index permissiveness | **DEFER** (diminishing-returns) — tracked as #305. |
| Nit | `ldr_width` function name used for STR dispatch | **DEFER** (style/naming) — folding into the cleanup tracked by #292. |
| Nit | Memory seed doesn't guide base registers into seeded region | **REJECT** (not-an-issue) — sound by design: SMT is the verification oracle for memory-bearing windows and `fast_only` is force-disabled, so the seeded region is informational rather than load-bearing. ADR-0007 covers this. |

Net new tracking issues filed in this iteration: #302 (sibling of #295), #303, #304, #305. Two REJECTs (#1 and the memory-seed nit) — happy to re-evaluate if you disagree.

_This assessment was made by two independent AI reviewers (Claude Opus 4.7 and Codex GPT-5.5)._

File: src/ir/instructions.rs [hunk 19]
```diff
@@ -940,8 +1100,159 @@ impl Instruction {
...
+    if address_base(addr) == Register::XZR {
+        return false;
+    }
+    // LDP rejects rt1 == rt2 (UNPREDICTABLE per ARM ARM). STP allows it —
+    // stores the same value twice — so this rule fires only for the
+    // load-pair caller.
+    if signed && width != AccessWidth::Word {
+        // LDPSW is the only "signed pair" form; it is always 32→64.
+        return false;
+    }
```
[the remark this answers is not in the document]
Comment (pmatos) [reply src/ir/instructions.rs:1208]: Fixed in `e7a76c3`.

Changed: `is_encodable_pair` in `src/ir/instructions.rs` now also rejects `width = Byte` / `Half` for the unsigned LDP/STP path. LDP/STP only have Word and Extended forms architecturally — there is no LDPB / LDPH / STPB / STPH — so Byte/Half pairs constructed by the IR or stochastic mutator are blocked at the encodability gate before reaching the assembler.

Validation: 2 new tests in `src/ir/instructions.rs` — `pair_byte_width_rejected_at_encodability` and `pair_half_width_rejected_at_encodability`.

_This assessment was made by two independent AI reviewers (Claude Opus 4.7 and Codex GPT-5.5). If you disagree, please reply and we'll re-evaluate._

File: src/parser/mod.rs [hunk 20]
```diff
@@ -1401,6 +1706,76 @@ pub fn parse_line(line: &str) -> Result<LineResult, ParseLineError> {
...
+            &operands,
+            crate::ir::types::AccessWidth::Half,
+            |rt, addr, w| Instruction::Ldr { rt, addr, width: w },
+        )
+        .map_err(ParseLineError::Other)?,
+        "ldrsb" => parse_single_reg_mem(
+            "ldrsb",
+            &operands,
+            crate::ir::types::AccessWidth::Byte,
+            |rt, addr, w| Instruction::Ldrs { rt, addr, width: w },
```
[the remark this answers is not in the document]
Comment (pmatos) [reply src/parser/mod.rs:1732]: **Deferred (correctness — high priority)** — Verified the claim: parser maps both `w0` and `x0` to `Register::X0` (via `parse_w_or_x_register`), and `sign_extend_load` always extends to 64 bits, so `ldrsb wN` would silently produce the X-form result. Same applies to `ldrsh wN` and (parsing-only) `ldrsw wN`.

Currently dormant — `grep -rn "ldrsb w\|ldrsh w\|ldrsw w"` in `src/` and `tests/` returns nothing, and no Capstone-tripwire row exercises a W-form signed load. So no current input miscompiles, but any future binary that disassembles to `ldrsb wN` would silently get the wrong register value.

The proper fix is non-trivial — either reject W-form for LDRS variants at the parser (surgical), or add a destination-width bit to the IR and thread it through concrete + SMT + assembler (proper). Tracked as **#306** so the next iteration picks it up before any real input surfaces the divergence.

_This assessment was made by two independent AI reviewers (Claude Opus 4.7 and Codex GPT-5.5). If you disagree, please reply and we'll re-evaluate._

File: src/ir/instructions.rs [hunk 21]
```diff
@@ -940,8 +1100,166 @@ impl Instruction {
...
+        AddressOperand::Imm { .. } => {}
+    }
+    if width == AccessWidth::Extended {
+        // The caller decides whether Extended is valid (LDR allows it, but
+        // is_encodable_ldr_like is also used for LDRS where it isn't). The
+        // Ldrs arm of is_encodable_aarch64 rejects Extended explicitly
+        // below; this helper conservatively allows it and the variant arm
+        // layers on the narrower rule.
+    }
+    true
```
[the remark this answers is not in the document]
Comment (pmatos) [reply src/ir/instructions.rs:1175]: **Deferred (diminishing returns)** — Verified gap. Same class as the LDP/STP imm-range check already landed in `6a2c43a`, but for single-register ops. The fix is mechanical (bound the immediate against the access-width-scaled band) but warrants its own test sweep across all 5 addressing modes × 4 access widths, and accepted forms today still route correctly through dynasm — out-of-range immediates panic at emit rather than miscompile. Tracked as #307 so it lands alongside the existing imm-range work.

_This assessment was made by two independent AI reviewers (Claude Opus 4.7 and Codex GPT-5.5). If you disagree, please reply and we'll re-evaluate._
